qa/suites/orch/cephadm/smoke-roleless/2-services/nfs-keepalive-only.yaml

```diff
@@ -0,0 +1,55 @@
+tasks:
+- vip:
+
+# make sure cephadm notices the new IP
+- cephadm.shell:
+    host.a:
+      - ceph orch device ls --refresh
+
+# stop kernel nfs server, if running
+- vip.exec:
+    all-hosts:
+      - systemctl stop nfs-server
+
+- cephadm.shell:
+    host.a:
+      - ceph fs volume create foofs
+
+# deploy nfs + keepalive-only ingress service
+- cephadm.apply:
+    specs:
+      - service_type: nfs
+        service_id: foo
+        placement:
+          count: 1
+        spec:
+          port: 2049
+          virtual_ip: "{{VIP0}}"
+      - service_type: ingress
+        service_id: nfs.foo
+        placement:
+          count: 1
+        spec:
+          backend_service: nfs.foo
+          monitor_port: 9002
+          virtual_ip: "{{VIP0}}/{{VIPPREFIXLEN}}"
+          keepalive_only: true
+- cephadm.wait_for_service:
+    service: nfs.foo
+- cephadm.wait_for_service:
+    service: ingress.nfs.foo
+
+# export and mount
+- cephadm.shell:
+    host.a:
+      - ceph nfs export create cephfs --fsname foofs --cluster-id foo --pseudo-path /fake
+
+# make sure mount can be reached over VIP, ensuring both that
+# keepalived is maintaining the VIP and that the nfs has bound to it
+- vip.exec:
+    host.a:
+      - mkdir /mnt/foo
+      - sleep 5
+      - mount -t nfs {{VIP0}}:/fake /mnt/foo
```

Comment:
```
023-03-22T14:45:45.339 DEBUG:teuthology.orchestra.run:got remote process result: 32
2023-03-22T14:45:45.340 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_teuthology_5c9057f4b1ecae7b7410e418e40b739ff0616d25/teuthology/run_tasks.py", line 103, in run_tasks
    manager = run_one_task(taskname, ctx=ctx, config=config)
  File "/home/teuthworker/src/git.ceph.com_teuthology_5c9057f4b1ecae7b7410e418e40b739ff0616d25/teuthology/run_tasks.py", line 82, in run_one_task
    return task(**kwargs)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_6f0f906272c8fb053c785bf0340c636123037ba1/qa/tasks/vip.py", line 60, in exec
    remote.run(
  File "/home/teuthworker/src/git.ceph.com_teuthology_5c9057f4b1ecae7b7410e418e40b739ff0616d25/teuthology/orchestra/remote.py", line 525, in run
    r = self._runner(client=self.ssh, name=self.shortname, **kwargs)
  File "/home/teuthworker/src/git.ceph.com_teuthology_5c9057f4b1ecae7b7410e418e40b739ff0616d25/teuthology/orchestra/run.py", line 455, in run
    r.wait()
  File "/home/teuthworker/src/git.ceph.com_teuthology_5c9057f4b1ecae7b7410e418e40b739ff0616d25/teuthology/orchestra/run.py", line 161, in wait
    self._raise_for_status()
  File "/home/teuthworker/src/git.ceph.com_teuthology_5c9057f4b1ecae7b7410e418e40b739ff0616d25/teuthology/orchestra/run.py", line 181, in _raise_for_status
    raise CommandFailedError(
teuthology.exceptions.CommandFailedError: Command failed on smithi019 with status 32: "sudo TESTDIR=/home/ubuntu/cephtest bash -ex -c 'mount -t nfs 10.0.31.19:/fake /mnt/foo'"
```
This mount didn't work. Thinking a mistake in the test procedure or which IP I used here, but will need to verify

Reply:
https://pulpito.ceph.com/adking-2023-03-15_01:34:51-orch:cephadm-wip-adk-testing-2023-03-14-1905-distro-default-smithi/7208435

Reply:
I tried to read the log. I am not clear what it means, but this could be relevant:
```
2023-03-22T14:45:45.336 INFO:teuthology.orchestra.run.smithi019.stderr:mount.nfs: Connection refused
```

Reply:
went looking a little bit. It seems the correct IP is getting set in the ganesha conf to bind to, but the bind is still failing.
Conf:
```
[root@smithi027 ~]# cat /var/lib/ceph/8e5b48e4-cda4-11ed-9afe-001a4aab830c/nfs.foo.0.0.smithi027.mjtihm/etc/ganesha/ganesha.conf 
# This file is generated by cephadm.
NFS_CORE_PARAM {
        Enable_NLM = false;
        Enable_RQUOTA = false;
        Protocols = 4;
        NFS_Port = 2049;
        Bind_addr = 10.0.31.27;
}

NFSv4 {
        Delegations = false;
        RecoveryBackend = 'rados_cluster';
        Minor_Versions = 1, 2;
}

RADOS_KV {
        UserId = "nfs.foo.0.0.smithi027.mjtihm";
        nodeid = "nfs.foo.0";
        pool = ".nfs";
        namespace = "foo";
}

RADOS_URLS {
        UserId = "nfs.foo.0.0.smithi027.mjtihm";
        watch_url = "rados://.nfs/foo/conf-nfs.foo";
}

RGW {
        cluster = "ceph";
        name = "client.nfs.foo.0.0.smithi027.mjtihm-rgw";
}

%url    rados://.nfs/foo/conf-nfs.foo
```
and the VIP 10.0.31.27 is reachable
```
root@smithi027 ~]# ping 10.0.31.27
PING 10.0.31.27 (10.0.31.27) 56(84) bytes of data.
64 bytes from 10.0.31.27: icmp_seq=1 ttl=64 time=0.124 ms
64 bytes from 10.0.31.27: icmp_seq=2 ttl=64 time=0.137 ms
64 bytes from 10.0.31.27: icmp_seq=3 ttl=64 time=0.091 ms
64 bytes from 10.0.31.27: icmp_seq=4 ttl=64 time=0.134 ms
^C
--- 10.0.31.27 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3061ms
rtt min/avg/max/mdev = 0.091/0.121/0.137/0.021 ms
```
but nfs can't bind to the port
```
Mar 28 20:25:44 smithi027 ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c-nfs-foo-0-0-smithi027-mjtihm[145428]: 28/03/2023 20:25:44 : epoch 64234d3b : smithi027 : ganesha.nfsd-7[main] lower_my_caps :NFS STARTUP :EVENT :currently set capabilities are: cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_setfcap=ep
Mar 28 20:25:44 smithi027 ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c-nfs-foo-0-0-smithi027-mjtihm[145428]: 28/03/2023 20:25:44 : epoch 64234d3b : smithi027 : ganesha.nfsd-7[main] gsh_dbus_pkginit :DBUS :CRIT :dbus_bus_get failed (Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory)
Mar 28 20:25:44 smithi027 ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c-nfs-foo-0-0-smithi027-mjtihm[145428]: 28/03/2023 20:25:44 : epoch 64234d3b : smithi027 : ganesha.nfsd-7[main] gsh_dbus_register_path :DBUS :CRIT :dbus_connection_register_object_path called with no DBUS connection
Mar 28 20:25:44 smithi027 ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c-nfs-foo-0-0-smithi027-mjtihm[145428]: 28/03/2023 20:25:44 : epoch 64234d3b : smithi027 : ganesha.nfsd-7[main] gsh_dbus_register_path :DBUS :CRIT :dbus_connection_register_object_path called with no DBUS connection
Mar 28 20:25:44 smithi027 ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c-nfs-foo-0-0-smithi027-mjtihm[145428]: 28/03/2023 20:25:44 : epoch 64234d3b : smithi027 : ganesha.nfsd-7[main] gsh_dbus_register_path :DBUS :CRIT :dbus_connection_register_object_path called with no DBUS connection
Mar 28 20:25:44 smithi027 ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c-nfs-foo-0-0-smithi027-mjtihm[145428]: 28/03/2023 20:25:44 : epoch 64234d3b : smithi027 : ganesha.nfsd-7[main] Bind_sockets_V6 :DISP :WARN :Cannot bind NFS udp6 socket, error 99 (Cannot assign requested address)
Mar 28 20:25:44 smithi027 ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c-nfs-foo-0-0-smithi027-mjtihm[145428]: 28/03/2023 20:25:44 : epoch 64234d3b : smithi027 : ganesha.nfsd-7[main] Bind_sockets :DISP :FATAL :Error binding to V6 interface. Cannot continue.
Mar 28 20:25:44 smithi027 systemd[1]: ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c@nfs.foo.0.0.smithi027.mjtihm.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Mar 28 20:25:44 smithi027 systemd[1]: ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c@nfs.foo.0.0.smithi027.mjtihm.service: Failed with result 'exit-code'.
Mar 28 20:25:55 smithi027 systemd[1]: ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c@nfs.foo.0.0.smithi027.mjtihm.service: Service RestartSec=10s expired, scheduling restart.
Mar 28 20:25:55 smithi027 systemd[1]: ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c@nfs.foo.0.0.smithi027.mjtihm.service: Scheduled restart job, restart counter is at 5.
Mar 28 20:25:55 smithi027 systemd[1]: Stopped Ceph nfs.foo.0.0.smithi027.mjtihm for 8e5b48e4-cda4-11ed-9afe-001a4aab830c.
Mar 28 20:25:55 smithi027 systemd[1]: ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c@nfs.foo.0.0.smithi027.mjtihm.service: Start request repeated too quickly.
Mar 28 20:25:55 smithi027 systemd[1]: ceph-8e5b48e4-cda4-11ed-9afe-001a4aab830c@nfs.foo.0.0.smithi027.mjtihm.service: Failed with result 'exit-code'.
Mar 28 20:25:55 smithi027 systemd[1]: Failed to start Ceph nfs.foo.0.0.smithi027.mjtihm for 8e5b48e4-cda4-11ed-9afe-001a4aab830c.
```
I was sure this had worked earlier, will need to try and find out what's broken now.

Reply:
Interestingly,  this works totally fine on my VMs. The NFS seems to start up okay
```
Mar 28 20:57:20 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:20 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_Init_admin_thread :NFS CB :EVENT :Admin thread initialized
Mar 28 20:57:20 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:20 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_rpc_cb_init_ccache :NFS STARTUP :EVENT :Callback creds directory (/var/run/ganesha) already exists
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] find_keytab_entry :NFS CB :WARN :Configuration file does not specify default realm while getting default realm name
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] gssd_refresh_krb5_machine_credential :NFS CB :CRIT :ERROR: gssd_refresh_krb5_machine_credential: no usable keytab entry found in keytab /etc/krb5.keytab for connection with host localhost
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_rpc_cb_init_ccache :NFS STARTUP :WARN :gssd_refresh_krb5_machine_credential failed (-1765328160:0)
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_Start_threads :THREAD :EVENT :Starting delayed executor.
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_Start_threads :THREAD :EVENT :gsh_dbusthread was started successfully
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_Start_threads :THREAD :EVENT :admin thread was started successfully
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_Start_threads :THREAD :EVENT :reaper thread was started successfully
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_Start_threads :THREAD :EVENT :General fridge was started successfully
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_start :NFS STARTUP :EVENT :-------------------------------------------------
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_start :NFS STARTUP :EVENT :             NFS SERVER INITIALIZED
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[main] nfs_start :NFS STARTUP :EVENT :-------------------------------------------------
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[dbus] gsh_dbus_thread :DBUS :CRIT :DBUS not initialized, service thread exiting
Mar 28 20:57:24 vm-02 ceph-f7c45226-cda9-11ed-9c62-52540087101d-nfs-foo-0-0-vm-02-zhqtkr[36922]: 28/03/2023 20:57:24 : epoch 642354a4 : vm-02 : ganesha.nfsd-6[dbus] gsh_dbus_thread :DBUS :EVENT :shutdown
```
and the mount works fine across multiple hosts
```
[root@vm-00 ~]# mount -t nfs 192.168.122.100:/fake /mnt/foo
[root@vm-00 ~]# 
[root@vm-00 ~]# 
[root@vm-00 ~]# touch /mnt/foo/test.txt
[root@vm-00 ~]# echo "HHH" > /mnt/foo/test.txt 
[root@vm-00 ~]# cat /mnt/foo/test.txt 
HHH
[root@vm-00 ~]# ssh root@vm-01
The authenticity of host 'vm-01 (192.168.122.155)' can't be established.
ECDSA key fingerprint is SHA256:3FNCSU9M5vpWeNK2o2wW4ejWd9qrA+k/+KzzdM7Uz64.
Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
Warning: Permanently added 'vm-01' (ECDSA) to the list of known hosts.
Activate the web console with: systemctl enable --now cockpit.socket

[root@vm-01 ~]# mkdir -p /mnt/foo
[root@vm-01 ~]# mount -t nfs 192.168.122.100:/fake /mnt/foo
[root@vm-01 ~]# cat /mnt/foo/test.txt 
HHH
[root@vm-01 ~]# 
```

Reply:
I wonder if it has something to do with IPv6? I think I remember that IPv6 doesn't work between nodes in the sepia lab? Not totally sure that's true, but maybe somewhere to start looking (even though the VIP is an IPv4 address). There was that one  `:WARN :Cannot bind NFS udp6 socket, error 99 (Cannot assign requested address)` line in the nfs journal from the failed test that maybe links to that.

Reply:
If that is the case you could try passing mount option `proto=tcp` to exclude tcp6 support. Maybe also include `vers=4` to try and force nfs 4 support?

Reply:
this turned out to be an issue with the VIP not being available for the NFS to bind to as keepalived and the NFS were placed on different nodes. Using just keepalived vs. keepalived + haproxy creates an additional requirement that the keepalived placement must match the placement of the backend service in order for the daemon in the backend service to bind to the VIP. Going to need to implement that linking between the two placements in this PR for things to work.

Reply:
Made a first pass attempt at this, although it felt a bit uncomfortable to implement. Will see if it at least gets the test to pass.

Reply:
```
2023-05-01T19:14:47.317164+0000 mgr.smithi134.mkferm (mgr.14221) 159 : cephadm [INF] Saving service mds.foofs spec with placement count:2
2023-05-01T19:14:49.982525+0000 mgr.smithi134.mkferm (mgr.14221) 163 : cephadm [INF] Saving service nfs.foo spec with placement count:1
2023-05-01T19:14:49.986773+0000 mgr.smithi134.mkferm (mgr.14221) 164 : cephadm [INF] Saving service ingress.nfs.foo spec with placement count:1
2023-05-01T19:14:53.273914+0000 mgr.smithi134.mkferm (mgr.14221) 167 : cephadm [INF] Deploying daemon mds.foofs.smithi159.bpmjzx on smithi159
2023-05-01T19:14:55.454643+0000 mgr.smithi134.mkferm (mgr.14221) 169 : cephadm [INF] Deploying daemon mds.foofs.smithi134.ubtlct on smithi134
2023-05-01T19:14:58.190095+0000 mgr.smithi134.mkferm (mgr.14221) 173 : cephadm [INF] Creating key for client.nfs.foo.0.0.smithi134.opxwfr
2023-05-01T19:14:58.194597+0000 mgr.smithi134.mkferm (mgr.14221) 174 : cephadm [INF] Ensuring nfs.foo.0 is in the ganesha grace table
2023-05-01T19:14:58.240400+0000 mgr.smithi134.mkferm (mgr.14221) 175 : cephadm [WRN] ganesha-rados-grace tool failed: rados_pool_create: -1
Can't connect to cluster: -1
terminate called after throwing an instance of 'std::bad_variant_access'
  what():  std::get: wrong index for variant

2023-05-01T19:14:58.245047+0000 mgr.smithi134.mkferm (mgr.14221) 176 : cephadm [ERR] Failed while placing nfs.foo.0.0.smithi134.opxwfr on smithi134: grace tool failed: rados_pool_create: -1
Can't connect to cluster: -1
terminate called after throwing an instance of 'std::bad_variant_access'
  what():  std::get: wrong index for variant

2023-05-01T19:15:00.277786+0000 mgr.smithi134.mkferm (mgr.14221) 179 : cephadm [INF] 10.0.31.134 is in 10.0.0.0/16 on smithi159 interface enp3s0f1
2023-05-01T19:15:00.282542+0000 mgr.smithi134.mkferm (mgr.14221) 180 : cephadm [INF] Deploying daemon keepalived.nfs.foo.smithi159.zclltz on smithi159
2023-05-01T19:15:17.445710+0000 mgr.smithi134.mkferm (mgr.14221) 195 : cephadm [INF] Fencing old nfs.foo.0.0.smithi134.opxwfr
2023-05-01T19:15:17.454136+0000 mgr.smithi134.mkferm (mgr.14221) 196 : cephadm [INF] Creating key for client.nfs.foo.0.1.smithi134.edymdd
2023-05-01T19:15:17.457479+0000 mgr.smithi134.mkferm (mgr.14221) 197 : cephadm [INF] Ensuring nfs.foo.0 is in the ganesha grace table
2023-05-01T19:15:17.565048+0000 mgr.smithi134.mkferm (mgr.14221) 198 : cephadm [INF] Creating rados config object: conf-nfs.foo
2023-05-01T19:15:17.613343+0000 mgr.smithi134.mkferm (mgr.14221) 199 : cephadm [INF] Creating key for client.nfs.foo.0.1.smithi134.edymdd-rgw
2023-05-01T19:15:17.623788+0000 mgr.smithi134.mkferm (mgr.14221) 200 : cephadm [INF] Deploying daemon nfs.foo.0.1.smithi134.edymdd on smithi134
```